cmd/otel-allocator/target/discovery.go

```diff
@@ -57,8 +61,17 @@ func NewDiscoverer(log logr.Logger, manager *discovery.Manager, hook discoveryHo
 	}
 }
 
+// GetScrapeConfigs retrieves current jobs and return a new copy of a map
+// for use outside of discoverer.
 func (m *Discoverer) GetScrapeConfigs() map[string]*config.ScrapeConfig {
-	return m.jobToScrapeConfig
+	m.configMtx.RLock()
+	defer m.configMtx.RUnlock()
+
+	jobToScrapeConfigCopy := make(map[string]*config.ScrapeConfig, len(m.jobToScrapeConfig))
```

Comment:
We actually got rid of this knowing there is a risk of a race condition, optimizing for memory performance to avoid making a copy constantly. ([see this PR for more](https://github.com/open-telemetry/opentelemetry-operator/pull/1293#issue-1469943346)) Would it be possible to keep track of when the map changes to avoid allocating this memory on every call to the server?

Reply:
This call is made every 60s (configurable) from the collector, and if you have a pool of 10+ collectors the memory usage on the TA increases heavily. We need to find a better solution for this codepath overall, and I'm planning on talking about it at our next SIG meeting.

Reply:
Got it, then it is indeed a bit trickier. I'm thinking whether we could directly return hash for comparing so would copy map only if changes were detected. Alternatively, I'm wondering whether the scrape config could be periodically checked / updated instead of on every `ScrapeConfigHandler` invocation, but I might be missing more context as to whether this is a possible solution.

Either way, happy to discuss this at SIG and find a proper solution, as this is currently blocking us from using target allocator.